File: gtci/16-subsets/python3/8-structurally-unique-bsts.py
class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findUniqueTrees(self, n):
        if n == 0:
            return []

        memo = {}

        def dfs(start, end):
            # base case -- no values remaining
            if start > end:
                return [None]

            # base case -- memoised
            if (start, end) in memo:
                return memo[(start, end)]

            trees = []

            # we fix one num to be the root
            # then the lower nums are the left subtree
            # and the greater nums are the right subtree
            for root_val in range(start, end + 1):
                left_subtrees = dfs(start, root_val - 1)
                right_subtrees = dfs(root_val + 1, end)

                for left in left_subtrees:
                    for right in right_subtrees:
                        root = TreeNode(root_val, left, right)

                        trees.append(root)

            memo[(start, end)] = trees
            return trees

        return dfs(1, n)


if __name__ == "__main__":
    sol = Solution()

    # n = 0 should return []
    assert sol.findUniqueTrees(0) == []

    # n = 1 should return 1 tree
    trees_1 = sol.findUniqueTrees(1)
    assert len(trees_1) == 1
    assert trees_1[0] is not None and trees_1[0].val == 1

    # Helper to serialize tree structure
    def serialize(root):
        if not root:
            return None
        return (root.val, serialize(root.left), serialize(root.right))

    # n = 2 should return 2 trees
    trees_2 = sol.findUniqueTrees(2)
    assert len(trees_2) == 2
    expected_2 = [
        (1, None, (2, None, None)),  # 1 as root, 2 as right child
        (2, (1, None, None), None),  # 2 as root, 1 as left child
    ]
    serialized_2 = [serialize(t) for t in trees_2]
    for e in expected_2:
        assert e in serialized_2

    # n = 3 should return 5 trees
    trees_3 = sol.findUniqueTrees(3)
    assert len(trees_3) == 5
    expected_3 = [
        (1, None, (2, None, (3, None, None))),
        (1, None, (3, (2, None, None), None)),
        (2, (1, None, None), (3, None, None)),
        (3, (1, None, (2, None, None)), None),
        (3, (2, (1, None, None), None), None),
    ]
    serialized_3 = [serialize(t) for t in trees_3]
    for e in expected_3:
        assert e in serialized_3
    trees_3 = sol.findUniqueTrees(3)
    assert len(trees_3) == 5

    print("All test cases passed.")

File: gtci/16-subsets/cpp/8-structurally-unique-bsts.cpp
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int x, TreeNode* left = nullptr, TreeNode* right = nullptr) : val(x), left(left), right(right) {}

    ~TreeNode() {
        delete left;
        delete right;
    }
};

class Solution {
public:
    struct pair_hash {
        template <typename T1, typename T2>
        std::size_t operator()(const std::pair<T1, T2>& p) const {
            return std::hash<T1>{}(p.first) ^ (std::hash<T2>{}(p.second) << 1);
        }
    };

    std::vector<TreeNode*> findUniqueTrees(int n) {
        if (n == 0)
            return {};

        std::unordered_map<std::pair<int, int>, std::vector<TreeNode*>, pair_hash> memo;

        return dfs(memo, 1, n);
    }

private:
    std::vector<TreeNode*> dfs(std::unordered_map<std::pair<int, int>, std::vector<TreeNode*>, pair_hash>& memo, const int& start, const int& end) {
        // base case -- no values remaining
        if (start > end)
            return { nullptr };
        
        // base case -- memoised
        if (memo.contains({start, end}))
            return memo[{start, end}];
        
        std::vector<TreeNode*> trees;
        for (int root_val = start; root_val < end + 1; root_val++) {
            std::vector<TreeNode*> leftSubtrees = dfs(memo, start, root_val - 1);
            std::vector<TreeNode*> rightSubtrees = dfs(memo, root_val + 1, end);

            for (auto left : leftSubtrees) {
                for (auto right : rightSubtrees) {
                    TreeNode* root = new TreeNode(root_val, left, right);

                    trees.push_back(root);
                }
            }
        }

        memo[{start, end}] = trees;
        return trees;
    }
};

int main() {
    Solution sol;

    // n = 0 should return 0 trees
    auto trees0 = sol.findUniqueTrees(0);
    assert(trees0.size() == 0);

    // n = 1 should return 1 tree
    auto trees1 = sol.findUniqueTrees(1);
    assert(trees1.size() == 1);

    // n = 2 should return 2 trees
    auto trees2 = sol.findUniqueTrees(2);
    assert(trees2.size() == 2);

    // n = 3 should return 5 trees
    auto trees3 = sol.findUniqueTrees(3);
    assert(trees3.size() == 5);

    std::cout << "All assertions passed." << std::endl;
    return 0;
}
